### DISCUSSION TITLE
Connecting to Azure SQL with access tokens: update

### DISCUSSION THREAD

**Discussion by jdmargulici** [sqlalchemy/sqlalchemy#11806] [Ideas]
The documentation provides a code sample for [connecting to a database on Azure with access tokens](https://docs.sqlalchemy.org/en/20/dialects/mssql.html#connecting-to-databases-with-access-tokens) using Active Directory -which is now "Entra". However the code is slightly out of date. I was able to make a working modification after some research and wanted to share it.

```
import struct
from sqlalchemy import create_engine, event
from sqlalchemy.engine.url import URL
from azure import identity


azure_credentials = identity.DefaultAzureCredential(exclude_interactive_browser_credential=False)
SQL_COPT_SS_ACCESS_TOKEN = 1256  # Connection option for access tokens, as defined in msodbcsql.h
TOKEN_URL = "https://database.windows.net/"  # The token URL for any Azure SQL database

# The connection url is what is outdated in the current documentation
connection_string = "Driver={ODBC Driver 17 for SQL Server};Server=tcp:[my-server].database.windows.net,1433;Database=[my-database];Encrypt=yes;TrustServerCertificate=no;Connection Timeout=30"
connection_url = URL.create("mssql+pyodbc", query={"odbc_connect": connection_string})
engine = create_engine(connection_url)

@event.listens_for(engine, "do_connect")
def provide_token(dialect, conn_rec, cargs, cparams):
    # remove the "Trusted_Connection" parameter that SQLAlchemy adds
    cargs[0] = cargs[0].replace(";Trusted_Connection=Yes", "")

    # create token credential  
    token_bytes = azure_credentials.get_token("https://database.windows.net/.default").token.encode("UTF-16-LE")
    token_struct = struct.pack(f'<I{len(token_bytes)}s', len(token_bytes), token_bytes)

    # apply it to keyword arguments
    cparams["attrs_before"] = {SQL_COPT_SS_ACCESS_TOKEN: token_struct}
```

**Comment by CaselIT** [comment]
Hi,

If you can send a pr to update the example it would be great!

**Reply by jdmargulici** [reply]
I really wish I could do this but I've already spent 20 minutes trying to understand the process I need to follow and it's just too taxing. For one there is no issue category that matches this topic, and second I don't want to have to clone the whole repo so that I can create a new branch and PR.

**Reply by CaselIT** [reply]
> I really wish I could do this but I've already spent 20 minutes trying to understand the process I need to follow and it's just too taxing. 

I'm sorry you found it too taxing, sqlalchemy uses the plain old github contribution model. make a fork, do changes, send a PR.

Thanks in any case
